### ISSUE TITLE
Compound matcher diff problem

### ISSUE THREAD

**Issue by myronmarston** [rspec/rspec-expectations#712]
Consider this spec:

``` ruby
RSpec::Matchers.define_negated_matcher :exclude, :include

RSpec.describe "Combining include and exclude" do
  let(:multiline_string) do
    <<-EOS
    Foo 1
    Bar 2
    Bazz 3
    EOS
  end

  example "include().and exclude()" do
    expect(multiline_string).to include("    Bar 3").and exclude("    Bar 4")
  end

  example "separate expectations" do
    expect(multiline_string).to include("    Bar 3")
    expect(multiline_string).to exclude("    Bar 4")
  end
end
```

Both specs fail (as is intentional) but notice the difference in the failures:

```
Failures:

  1) Combining include and exclude include().and exclude()
     Failure/Error: expect(multiline_string).to include("    Bar 3").and exclude("    Bar 4")
       expected "    Foo 1\n    Bar 2\n    Bazz 3\n" to include "    Bar 3"
     # ./spec/compound_diff_problem_spec.rb:13:in `block (2 levels) in <top (required)>'

  2) Combining include and exclude separate expectations
     Failure/Error: expect(multiline_string).to include("    Bar 3")
       expected "    Foo 1\n    Bar 2\n    Bazz 3\n" to include "    Bar 3"
       Diff:
       @@ -1,2 +1,4 @@
       -    Bar 3
       +    Foo 1
       +    Bar 2
       +    Bazz 3

     # ./spec/compound_diff_problem_spec.rb:17:in `block (2 levels) in <top (required)>'

Finished in 0.00369 seconds (files took 0.15202 seconds to load)
2 examples, 2 failures
```

You lose the diff when you make it a compound matcher expression.  This is unfortunate, as the diff is so useful.  We should find a way to preserve the diff.  Unfortunately, the diff is not produced by the matcher itself but instead by rspec-expectations by comparing `matcher.expected` to `matcher.actual`...but with a compound matcher there are multiple `expected` values (one per matcher in the compound expression) and there's not a simple way to diff them.  We really need it to diff each matcher individually, and then include that in the failure output...but diffing isn't a responsibility of matchers; matchers simply say they are diffable and expose expected/actual, and I don't particularly want to make diffing a responsibility of the matchers.

Anyone got an idea for how best to solve this?

/cc @rspec/rspec

**Comment by waterlink** [comment]
How is compound matcher is different from just having matchers separated in separate expectations? So I suggest just iterating over matchers in compound matcher and rendering which matcher and diff for each of them (unless diff is empty).

In case of `and` compound matcher, diff can be easily rendered for first failed matcher. For `or` diff-printing routine should be run on both sides of `or`.

**Comment by JonRowe** [comment]
It's different because it combines the failure messages, evidentially we don't have a way for handling the diff against multiple expected states.

**Comment by myronmarston** [comment]
Also, we specifically don't want to run the two matchers as sequentially distinct expectations (as you are suggesting).  To handle a case like this:

``` ruby
x, y = 0, 0
expect {
  x +=1
  y += 2
}.to change { x }.by(1).and change { y }.by(2)
```

...it's important that the `expect` block is _not_ evaluated twice (otherwise it would fail when it should pass).  The compound matchers present a matcher interface (which allows users to treat them just like any other matcher) while delegating to two matchers internally (either of which could be another compound matcher) in appropriate ways.

**Comment by waterlink** [comment]
Ah, so compound matcher is actually yet another matcher, and it is not desired to put diff handling logic there.

**Comment by waterlink** [comment]
So what about allowing `matcher.expected` to return either one or a list of expected values?

Or even having explicit `#multiple_expected` (or better name I cannot come up with yet) and in `ExpectationHelper::handle_failure` handle matchers with `#multiple_expected` by iterating over them and rendering multiple diffs.

In this case `#multiple_expected` will be something like this for compound matcher:
- if `and` compound matcher:
  - if first matcher does not match: return `#multiple_expected` for first matcher (if responds_to, otherwise return `[#expected]`)
  - if second matcher does not match: the same but for second matcher
  - `[]` otherwise
- if `or` compound matcher:
  - map failed matchers (of two of them), with flattening:
    - either `#multiple_expected` or `[#expected]` whichever it is respond_to

Still question about mixing in compound matcher diffable and non-diffable matcher stays open.

WDYT? @myronmarston @JonRowe

**Comment by myronmarston** [comment]
That's certainly one option I had mulled over in my mind, but I don't necessarily want to add another method to the matcher protocol (which `multiple_expected` would be) just for this one matcher.

We could potentially do the diff internally inside the compound matchers and include it in `failure_message`, but I don't like that either.

In short, there's no option that I really like yet.

**Comment by waterlink** [comment]
There is another option: make `#expected` return not a raw value, but object with behaviour, which will understand its expected value (or values), its actual value and how to present them (diffable or not), and even generate a diff. So basically shift responsibility of diffing to this object.

**Comment by myronmarston** [comment]
That'd be a breaking change -- `expected` is part of the publish matcher protocol that custom matchers may implement.

**Comment by waterlink** [comment]
It can be a breaking change unless we make it backward compatible. Something along the lines:

```ruby
ExpectedValue.coerce_from_matcher(matcher).failure_message
```

which will call `#expected` on `matcher` and
will return what it gets as a result, if it is already an `ExpectedValue`,
otherwise will wrap the value with new instance of `ExpectedValue` and gather some additional info from `matcher` (like `diffable?`, `actual`, etc).

In this case, if somebody returns some value (like String, Fixnum, or arbitary object or class), then it is fine since it will be wrapped in `ExpectedValue` with default behavior. But if the matcher is aware, it can return the `ExpectedValue` with desired behavior.

**Comment by JonRowe** [comment]
I think that's overcomplicating things slightly... Can we not figure out a way to make the differ away of compound expectations?

**Comment by waterlink** [comment]
Yes, it is possible to make diffing code aware of compound matchers. I tried it out, and turns out that it duplicates a bunch of logic of what is inside of `And#failure_message` and `Or#failure_message`, particularly conditions:

**Comment by waterlink** [comment]
So summarizing, we have 3 options:

- extend matcher protocol and make diffing a responsibility of `And` and `Or` matchers
- extend what `#expected` can return (a normal value, or wrapped value)
- extend diffing logic with compound aware code

If I think in terms of OO, I would definitely defy 3rd option, because it creates duplication of knowledge of how to treat different compound matchers, and would actually go for 1st (simpler, but extends matcher protocol) or 2nd (more complicated, but could be very elegant) option.

If I think in terms of avoiding adding additional stuff to matcher protocol (which could result in more broader public interface and hardly maintainable code), then I would shoot for 2nd or 3rd option (preferring 3rd one, because it is simpler).

And now I think, there is another option is to add private api for compound matchers only, that does what we want, but not add it to matchers protocol at all.

So, WDYT?

**Comment by myronmarston** [comment]
After thinking about this some more, I think I like the 2nd option best.  Initially I was going to suggest that we could just make `expected` return an array of values in this case, but that won't actually work properly; in non-compound matchers we support diffing an array against an array, and there's no way for the diffing logic to know that in the compound matcher case the given array should be used as the basis for multiple diffs (one per array element) rather than using the array as a single object to diff against.

So I think you're right that we have to make `expected` return a special "wrapper" type that the diffing logic can use to know that it needs to produce multiple diffs.  I'd suggest something like `ExpectedsForMultipleDiffs` as that name is very explicit about its purpose (much more so than `ExpectedValue`, suggested before).  Doing this wouldn't necessarily require us to update the matcher protocol -- given that I can't imagine an end-user matcher ever needing this behavior (as it's really specific to compound matchers) it could be an unpublished capability, although it's simple enough that we may just want to update the matcher protocol docs to mention it.  (I'm on the fence about that).

One other thing to consider is that `diffable?` on the compound matchers should be written to only return true if one of the failing matchers is itself diffable.

Do you want to take a stab at this, @waterlink?

**Comment by waterlink** [comment]
@myronmarston, yes I will take a stab at this, 2nd option also fine with me.

**Comment by myronmarston** [comment]
Addressed by #713.
